README.org
* org-z

#+begin_quote
Lightweight, Org-mode flavored zettelkasten links.
#+end_quote

org-z is an Emacs global minor mode that enables a lightweight, Org mode style zettelkasten. Unlike a traditional zettelkasten, org-z focuses on headings rather than pages, allowing you to make hyperlinks within a single Org mode document.

By treating headings, rather than pages, as first-class, we can use standard functions that ship with org like [[https://orgmode.org/manual/Handling-Links.html][org-store-link]] to maximum effect.

org-z has the following features:

 - Insert links to org headings with interactive search (~C-c-.~ by default). [[https://github.com/emacs-helm/helm][Helm]] and [[https://github.com/raxod502/selectrum][Selectrum]] are supported.
 - Links are managed automatically using [[https://orgmode.org/manual/Handling-Links.html][org-store-link]] and org-ids
 - Links to new / missing targets auto-create targets in a configurable file location (~new.org~ by default). Creation of the new heading is customizable using a capture template.
 - [[./CHANGELOG.org#backlinks][Backlinks]] 

#+CAPTION: Inserting a link to an existing heading
#+NAME:   existing-heading
[[./images/org-z-existing-heading.gif]]

#+CAPTION: Inserting a link to a new heading
#+NAME: new-heading
[[./images/org-z-new-heading.gif]]

** Installation

Install with straight.el:

#+begin_src emacs-lisp
  (use-package org-z
    :straight (org-z :type git :host github :repo "landakram/org-z")
    :config
    (org-z-mode 1))
#+end_src

org-z requires a completion backend. [[https://github.com/emacs-helm/helm][Helm]] and [[https://github.com/raxod502/selectrum][Selectrum]] are currently supported. You'll need to load one of the completion backend packages, in addition to org-z itself:

#+begin_src emacs-lisp
  (use-package org-z-selectrum
    :straight (org-z-selectrum :type git :host github :repo "landakram/org-z"))
#+end_src

Loading the package sets the completion backend by appending to ~org-z-completion-backends~ and setting ~org-z-completion-backend~.

** A lightweight zettelkasten

I combine org-z with a daily journal, implemented as a ~file+datetree~ capture template, and [[./CHANGELOG.org#backlinks][org-z's support for backlinks]].

#+begin_src emacs-lisp
  (setq org-capture-templates
        '(;; ...
          ("j" "Journal entry" entry (file+datetree "~/org/journal.org")
           "* %?\n" :unnarrowed t)))

  (use-package org
    :straight org-plus-contrib
    :config
    (require 'org-eldoc))

  (use-package org-z
    :straight (org-z :type git :host github :repo "landakram/org-z")
    :general
    (leader-def :infix "o"
      "b" 'org-z-backlinks-at-point)
    :config
    (org-z-mode 1))
#+end_src

As I go about my day, I make journal entries that link to various headings in my collection of org files. I make these links by invoking ~org-z-insert-link~ as I write my journal entry. I periodically inspect backlinks using ~org-z-backlinks-at-point~ to investigate connections. By enabling ~org-eldoc~, I get to see the # of backlinks for a given heading when I hover over it.

Every once in a while, I review my journal entries and ~org-refile~ headings or simply re-organize journal entry knowledge into "longer-term storage" (~reference.org~ or ~projects.org~).

** Motivation

I was inspired by [[https://github.com/org-roam/org-roam][org-roam]], but wanted something that fit my existing org workflow, which is a handful of files with deep headings rather than a file per topic.


CHANGELOG.org
* Changelog
** v0.0.5
*** Features
**** Backlinks
:PROPERTIES:
:ID:       a08e2a03-19be-46b0-96fc-43a6556e3710
:END:
 - Add a function ~org-z-backlinks-at-point~ that can be used to inspect backlinks to the current headline. This function uses completing-read to allow jumping through a backlink. This function is not bound to any key by default, leaving keybindings open to the user.

 - If org-eldoc is loaded, show the number of backlinks for the current headline at point. This functionality can be turned off by unsetting ~org-z-show-backlinks~.

   This function respects ~eldoc-echo-area-use-multiline-p~. If t, then the backlink message is always displayed on a new line. If truthy, message is displayed same line if there's space, or a new line if there is not enough space. If false, it is always displayed on single line (header path will be truncated).
** v0.0.4
*** Bug fixes
**** Fix org-z-selectrum
~selectrum-read~ was deprecated and removed from selectrum. Use ~selectrum--read~ instead.
** v0.0.3
*** Bug fixes
**** Fix completion backends
***** Fix missing imports
These were causing errors like: ~Symbol's function definition is void: org-ql-search-directories-files~.
***** Fix helm backend not showing results
Because ~org-z--format-org-ql-heading~ returned a cons cell, ~org-ql-select~ tried to flatten it and failed silently due to ~ignore-errors~. Return text with a point-marker property instead.
** v0.0.2
*** Breaking changes
**** Completion backends
org-z now supports both Helm and Selectrum for completion. The completion backend must be explicitly set by loading one of ~org-z-helm~ or ~org-z-selectrum~ (which each depend on a respective completion framework) in addition to ~org-z~. Working configuration looks like this:

#+begin_src emacs-lisp
  (use-package org-z
    :straight (org-z :type git :host github :repo "landakram/org-z")
    :config
    (org-z-mode 1))

  (use-package org-z-selectrum
    :straight (org-z-selectrum :type git :host github :repo "landakram/org-z"))
#+end_src
*** Features
**** Sort links with prescient.el
Results can be optionally sorted with [[https://github.com/raxod502/prescient.el][prescient.el]] by setting ~org-z-use-prescient~. Note that if you already have prescient enabled through your completion system, you should leave this as nil. 


test/helm/init.el
(setq user-init-file (or load-file-name (buffer-file-name)))
(setq user-emacs-directory (file-name-directory user-init-file))

(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(setq package-enable-at-startup nil)
(package-initialize)
(when (not package-archive-contents)
  (package-refresh-contents))

(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(straight-use-package 'use-package)

(use-package helm
  :straight t)

(package-install-file "org-z.el")
(package-install-file "org-z-helm.el")

(use-package org-z
  :config
  (org-z-mode 1))

(use-package org-z-helm)


test/straight/init.el
(setq user-init-file (or load-file-name (buffer-file-name)))
(setq user-emacs-directory (file-name-directory user-init-file))

(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(setq package-enable-at-startup nil)
(package-initialize)
(when (not package-archive-contents)
  (package-refresh-contents))

(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(straight-use-package 'use-package)

(use-package org-z
  :straight (org-z :type git :host github :repo "landakram/org-z")
  :config
  (org-z-mode 1))

(use-package org-z-selectrum
  :straight (org-z-selectrum :type git :host github :repo "landakram/org-z"))


test/selectrum/init.el
(setq user-init-file (or load-file-name (buffer-file-name)))
(setq user-emacs-directory (file-name-directory user-init-file))

(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(setq package-enable-at-startup nil)
(package-initialize)
(when (not package-archive-contents)
  (package-refresh-contents))

(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(straight-use-package 'use-package)

(use-package selectrum
  :straight t)

(package-install-file "org-z.el")
(package-install-file "org-z-selectrum.el")

(use-package org-z
  :config
  (org-z-mode 1))

(use-package org-z-selectrum)


org-z-helm.el
;;; org-z-helm.el --- Helm completion backend for org-z -*- coding: utf-8; lexical-binding: t; -*-

;; Copyright (C) 2020 Mark Hudnall <me@markhudnall.com>

;; Author: Mark Hudnall <me@markhudnall.com>
;; URL: https://github.com/landakram/org-z
;; Keywords: org-mode
;; Version: 0.0.3
;; Package-Requires: ((emacs "27.1") (org-z "0.0.2") (org-ql "0.6-pre") (helm "3.3") (helm-org "1.0") (helm-org-ql "0.6-pre"))
;; Keywords: org-mode, outlines

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; Helm completion backend for org-z. This is an extra package, since org-z.el doesn't depend
;; on a specific completion backend.

;;; Code:

(require 'cl-lib)
(require 'cl-generic)
(require 'org)
(require 'org-id)
(require 'org-capture)

(require 'org-z)

(require 'org-ql)
(require 'org-ql-search)
(require 'helm-org-ql)

(defun org-z-insert-link--fallback ()
  (helm-build-dummy-source "Create link to new heading"
    :action (helm-make-actions
             "Insert link to new heading"
             #'org-z--insert-link-to-new-heading)))

(defun org-z-helm-candidate-transformer (candidates _)
  (cond ((and org-z-use-prescient (fboundp 'prescient-sort-compare))
         (sort candidates (lambda (c1 c2)
                            (prescient-sort-compare (car c1) (car c2)))))
        (org-z-use-prescient
         (user-error "org-z-use-prescient is t but prescient isn't loaded. Make sure prescient is installed and required."))
        (t
         candidates)))

(defun org-z-helm-org-ql-source (buffers-files)
  (helm-make-source "org-z" 'helm-source-sync
    :candidates (lambda ()
                  (let* ((query (org-ql--query-string-to-sexp helm-pattern))
                         (window-width (window-width (helm-window))))
                    (when query
                      (with-current-buffer (helm-buffer-get)
                        (setq helm-org-ql-buffers-files buffers-files))
                      ;; Ignore errors that might be caused by partially typed queries.
                      (ignore-errors
                        (mapcar (lambda (candidate)
                                  (let ((point-marker (get-text-property 0 'point-marker candidate)))
                                    (cons candidate point-marker)))
                                (org-ql-select buffers-files query
                                  :action `(org-z--format-org-ql-heading ,window-width)))))))
    :match #'identity
    :filtered-candidate-transformer #'org-z-helm-candidate-transformer
    :fuzzy-match nil
    :multimatch nil
    :nohighlight t
    :volatile t
    :keymap helm-org-ql-map
    :action helm-org-ql-actions))

(defun org-z-helm-select-action-hook ()
  (let ((selection (helm-get-selection nil t)))
    (cond
     ((and org-z-use-prescient (fboundp 'prescient-remember))
      (prescient-remember selection))
     (org-z-use-prescient
      (user-error "org-z-use-prescient is t but prescient isn't loaded. Make sure prescient is installed and required.")))))

(cl-defstruct org-z--helm-backend)

(cl-defmethod org-z--insert-link ((_ org-z--helm-backend))
  (when (not (fboundp 'helm))
    (user-error "org-z-completion-backend is 'helm but helm isn't loaded. Make sure helm is installed and required."))
  (require 'helm-org-ql)

  (let* ((helm-candidate-separator " ")
         (_ (add-to-list 'helm-org-ql-actions '("Insert link" . org-z--insert-link-to-candidate)))
         (org-sources (list (org-z-helm-org-ql-source (org-ql-search-directories-files :directories org-z-directories)))))
    (add-hook 'helm-before-action-hook #'org-z-helm-select-action-hook)
    (helm
     :input (thing-at-point 'symbol 'no-properties)
     :sources (append org-sources (list (org-z-insert-link--fallback)))
     :buffer "*org-z-insert-link*")
    (pop helm-before-action-hook)
    (pop helm-org-ql-actions)))

(setq org-z-completion-backends (plist-put org-z-completion-backends 'helm (make-org-z--helm-backend)))
(setq org-z-completion-backend 'helm)

(provide 'org-z-helm)

;;; org-z-helm.el ends here


org-z-selectrum.el
;;; org-z-selectrum.el --- Selectrum completion backend for org-z -*- coding: utf-8; lexical-binding: t; -*-

;; Copyright (C) 2020 Mark Hudnall <me@markhudnall.com>

;; Author: Mark Hudnall <me@markhudnall.com>
;; URL: https://github.com/landakram/org-z
;; Version: 0.0.4
;; Package-Requires: ((emacs "27.1") (org-z "0.0.4") (org-ql "0.6-pre") (selectrum "3.0"))
;; Keywords: org-mode, outlines

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; Selectrum completion backend for org-z. This is an extra package, since org-z.el doesn't depend
;; on a specific completion backend.

;;; Code:

(require 'cl-lib)
(require 'cl-generic)
(require 'org)
(require 'org-id)
(require 'org-capture)

(require 'org-z)

(require 'org-ql)
(require 'org-ql-search)
(require 'selectrum)

(cl-defstruct org-z--selectrum-backend)

(cl-defmethod org-z--insert-link ((_ org-z--selectrum-backend))
  (when (not (fboundp 'selectrum--read))
    (user-error "org-z-completion-backend is 'selectrum but selectrum isn't loaded. Make sure selectrum is installed and required."))

  (let* ((buffers-files (org-ql-search-directories-files :directories org-z-directories))
         (candidate-fn (lambda (pattern)
                         (let* ((query (org-ql--query-string-to-sexp pattern))
                                (window-width (window-text-width)))
                           (when query
                             (ignore-errors
                               ;; Ignore errors that might be caused by partially typed queries.
                               (org-ql-select buffers-files query
                                 :action `(org-z--format-org-ql-heading ,window-width)))))))
         (result
          ;; This is sort of an ugly hack. selectrum--read refines candidates using `completion-styles', which filters
          ;; the org-ql query further by the input. We're already effectively handling our own filtering in the candidate-fn
          ;; so if the user has default selectrum (they are not using e.g. prescient), we ignore `selectrum-refine-candidates-function'
          ;; for the completion
          (let ((selectrum-refine-candidates-function
                 (if (eq selectrum-refine-candidates-function #'selectrum-refine-candidates-using-completions-styles)
                     (lambda (input cands) cands)
                   selectrum-refine-candidates-function)))
            (selectrum--read "Insert link: " candidate-fn
                             :initial-input (thing-at-point 'symbol 'no-properties)))))
    (if-let ((point-marker (get-text-property 0 'point-marker result)))
        (org-z--insert-link-to-candidate point-marker)
      (org-z--insert-link-to-new-heading result))))

(setq org-z-completion-backends (plist-put org-z-completion-backends 'selectrum (make-org-z--selectrum-backend)))
(setq org-z-completion-backend 'selectrum)

(provide 'org-z-selectrum)

;;; org-z-selectrum.el ends here


org-z.el
;;; org-z.el --- Lightweight, Org-mode flavored zettelkasten links -*- coding: utf-8; lexical-binding: t; -*-

;; Copyright (C) 2020 Mark Hudnall <me@markhudnall.com>

;; Author: Mark Hudnall <me@markhudnall.com>
;; URL: https://github.com/landakram/org-z
;; Keywords: org-mode
;; Version: 0.0.5
;; Package-Requires: ((emacs "27.1") (org "9.3") (dash "2.12") (f "0.18.1") (s "1.10.0"))
;; Keywords: org-mode, outlines

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; Utilities for linking in org-mode

;;; Code:

(require 'cl-lib)
(require 'cl-generic)
(require 'org)
(require 'org-id)
(require 'org-capture)

(require 'f)
(require 's)
(require 'dash)

(defgroup org-z nil
  "org-z customizable variables."
  :group 'org)

(defcustom org-z-directories (list org-directory)
  "Directories in which org-z will look for org files."
  :type 'list
  :group 'org-z)

(defcustom org-z-new-headings-file (f-join org-directory "new.org")
  "File in which to write new headings when inserting a link to a heading that does not already exist."
  :type 'file
  :group 'org-z)

(defcustom org-z-refile-missing-heading nil
  "After inserting a link to a missing heading, the missing heading can be refiled. Valid options are:

nil: No refiling will occur.
'interactive: The heading will be refiled interactively.
'subheading: The heading will be automatically refiled to a subheading of the current heading (where the link is being inserted).
'sibling: The heading will be automatically refiled to a sibling of the current heading (where the link is being inserted)."
  :type 'symbol
  :group 'org-z)

(defun org-z-capture--templates (heading)
  "The capture templates used by org-z to create a heading when inserting a link to a heading that doesn't exist."
  (let ((body (concat "* " heading "%?\n:LOGBOOK:\n- Added %U\n:END:\n")))
    `(("d" "default" entry (file org-z-new-headings-file)
       ,body
       :immediate-finish t))))

(defcustom org-z-capture-templates #'org-z-capture--templates
  "The capture templates used by org-z to create a heading when inserting a link to a
heading that doesn't exist. This is a function which returns a list of capture templates
using the same syntax as `org-capture-templates'. The function accepts a single string
argument, which is the missing heading."
  :type 'function
  :group 'org-z)

(defcustom org-z-use-prescient nil
  "When t, use prescient.el to sort results. If you already have prescient enabled through
your completion system, you should leave this as nil."
  :type 'boolean
  :group 'org-z)

(defcustom org-z-completion-backend nil
  "The completion backend to use when searching for links to insert. Valid options are 'helm or 'selectrum.

Note that completion backends are implemented in separate packages. You should ensure that a completion backend
package such as org-z-selectrum is installed in addition to org-z."
  :type '(choice
          (const :tag "Helm" helm)
          (const :tag "Selectrum" selectrum))
  :group 'org-z)

(defun org-z--store-link-to-candidate (candidate)
  "Store link to CANDIDATE."
  (let ((buffer (marker-buffer candidate))
        (pos (marker-position candidate)))
    (save-excursion
      (with-current-buffer buffer
        (goto-char pos)
        (call-interactively 'org-store-link)))))

(defun org-z--insert-link-to-candidate (candidate)
  "Insert link to CANDIDATE in current location."
  (interactive)
  (org-z--store-link-to-candidate candidate)
  (call-interactively 'org-insert-link))

(defvar org-z-insert-missing-after-hook nil
  "A hook that is called after inserting a link to a missing heading.")

(defun org-z-capture--before-finalize-hook ()
  "During org-z-capture, store a link to the just-inserted headline."
  (let ((buf (org-capture-get :buffer))
        (point (org-capture-get :insertion-point)))
    (with-current-buffer buf
      (save-excursion
        (save-restriction
          (widen)
          (goto-char point)
          (call-interactively 'org-store-link)))))
  (remove-hook 'org-capture-before-finalize-hook #'org-z-capture--before-finalize-hook))

(defun org-z-capture--after-finalize-hook ()
  "After org-z-capture, insert a link."
  (call-interactively 'org-insert-link)
  (remove-hook 'org-capture-after-finalize-hook #'org-z-capture--after-finalize-hook))

(defun org-z-insert-missing (heading)
  "When inserting a link to a headline that doesn't exist, use a custom capture template to add the new headline."
  (let ((org-capture-templates (funcall org-z-capture-templates heading))
        goto keys)
    (when (= (length org-capture-templates) 1)
      (setq keys (caar org-capture-templates)))
    (org-capture goto keys)))

(defun org-z--refile-missing ()
  (let ((refile-loc
         (list
          (org-display-outline-path t t nil t)
          (buffer-file-name (buffer-base-buffer))
          nil
          (org-with-wide-buffer
           (org-back-to-heading t)
           (point-marker)))))

    (org-mark-ring-push)

    (org-capture-goto-last-stored)

    (cond ((eq org-z-refile-missing-heading 'interactive)
           (org-refile))

          ((eq org-z-refile-missing-heading 'subheading)
           (org-refile nil nil refile-loc))

          ((eq org-z-refile-missing-heading 'sibling)
           (org-refile nil nil refile-loc)
           (org-refile-goto-last-stored)
           (outline-promote)))

    (org-mark-ring-goto)))

(defun org-z--insert-link-to-new-heading (candidate)
  (add-hook 'org-capture-before-finalize-hook #'org-z-capture--before-finalize-hook)
  (add-hook 'org-capture-after-finalize-hook #'org-z-capture--after-finalize-hook)
  (org-z-insert-missing candidate)

  (when org-z-refile-missing-heading
    (org-z--refile-missing))

  (run-hooks 'org-z-insert-missing-after-hook))

(defun org-z--format-org-ql-heading (window-width)
  "Return string for completion backend for heading at point.
WINDOW-WIDTH should be the width of the window."
  (font-lock-ensure (point-at-bol) (point-at-eol))
  (let* ((prefix (concat (buffer-name) ":"))
         (width (- window-width (length prefix)))
         (heading (org-get-heading t))
         (path (-> (org-get-outline-path)
                   (org-format-outline-path width nil "")
                   (org-split-string "")))
         (path (concat (s-join "/" path) "/" heading)))
    (propertize
     (concat prefix path)
     'point-marker
     (point-marker))))

(defun org-z--backlinks-query (marker)
  "Return an org-ql query to find backlinks for the heading at MARKER."
  (let* ((id (org-entry-get marker "ID"))
         (custom-id (org-entry-get marker "CUSTOM_ID"))
         ;; FIXME: Do CUSTOM_ID links also have an "id:" prefix?
         (query (cond ((and id custom-id)
                       ;; This will be slow because it isn't optimized to a single regexp.  :(
                       (warn "Entry has both ID and CUSTOM_ID set; query will be slow")
                       `(or (link :target ,(concat "id:" id))
                            (link :target ,(concat "id:" custom-id))))
                      ((or id custom-id)
                       `(link :target ,(concat "id:" (or id custom-id))))
                      (t (error "Entry has no ID nor CUSTOM_ID property")))))
    query))

(defun org-z-backlinks-at-point ()
  "Show backlinks for an org header at the current point using `completing-read'.
Selecting a backlink jumps to the backlink's heading."
  (interactive)
  (let* ((marker (point-marker))
         (query (org-z--backlinks-query marker))
         (buffers-files (org-ql-search-directories-files :directories org-z-directories))
         (window-width (window-text-width))
         (results (org-ql-select buffers-files query :action `(org-z--format-org-ql-heading ,window-width)))
         ;; completing-read strips all text-properties on the result, which is how we normally convey the point-marker.
         ;; Build up an alist that we can look up after the completing-read instead.
         (results-lookup (mapcar (lambda (r)
                                   `(,(substring-no-properties r 0) . ,(get-text-property 0 'point-marker r)))
                                 results)))
    (if (not results)
        (message "No backlinks.")
      (let ((backlink (completing-read "Backlinks: " results)))
        (setq org-z--debug-backlink backlink)
        (if-let* ((result (assoc backlink results-lookup))
                  (point-marker (cdr result))
                  (buf (marker-buffer point-marker)))
            (progn
              (push-mark)
              (switch-to-buffer buf)
              (goto-char point-marker))
          (message "No backlink selected."))))))

(cl-defstruct org-z--completion-backend
  org-z--insert-link)

(cl-defgeneric org-z--insert-link (org-z--completion-backend))

(defvar org-z-completion-backends '()
  "Completion backends for org-z. This is a plist where keys are `org-z-completion-backend' and
values are instances of an `org-z--completion-backend'.")

;;;###autoload
(defun org-z-insert-link (prefix)
  "Begin inserting a link to an org headline at point. A completion backend allows interactively searching for headlines to link."
  (interactive "P")
  (let* ((org-z-refile-missing-heading (cond ((equal prefix '(4))
                                              'sibling)
                                             ((equal prefix '(16))
                                              'subheading)
                                             (t
                                              org-z-refile-missing-heading)))
         (current-prefix-arg nil)
         (completion-backend (plist-get org-z-completion-backends org-z-completion-backend)))
    (when (not completion-backend)
      (user-error "Completion backend is nil! This means that either `org-z-completion-backend' is invalid, or you forgot to load one of the org-z completion backends, which are separate packages. If you use helm for example, make sure that org-z-helm is loaded in addntion to org-z."))

    (org-z--insert-link completion-backend)))

(defun org-z-knowledge--search (targets &optional rg-opts)
  (let ((helm-rg-default-extra-args rg-opts)
        (helm-rg-default-directory (expand-file-name "~/")))
    (helm-rg
     nil
     nil
     targets)))

(defcustom org-z-knowledge-dirs org-z-directories
  "Directories in which to perform full-text knowledge search."
  :type 'list
  :group 'org-z)

(defcustom org-z-knowledge-filetypes (list "org" "md")
  "File-types in which to perform full-text knowledge search."
  :type 'list
  :group 'org-z)

(defcustom org-z-show-backlinks t
  "When t, show backlinks for the entry at point in the echo area. This functionality requires org-eldoc."
  :type 'boolean
  :group 'org-z)

;;;###autoload
(defun org-z-knowledge-search ()
  "Perform full-text search on matching files in `org-z-knowledge-dirs'."
  (interactive)
  (let ((rg-opts (flatten-list (mapcar
                                (lambda (ft)
                                  `("-t" ,ft))
                                org-z-knowledge-filetypes))))
    (org-z-knowledge--search org-z-knowledge-dirs rg-opts)))

(defun org-z-eldoc-get-breadcrumb ()
  (let* ((marker (point-marker))
         (buffers-files (org-ql-search-directories-files :directories org-z-directories))
         (query (ignore-errors
                  (org-z--backlinks-query marker)))
         (window-width (window-text-width))
         (results (ignore-errors
                    (and query
                         (org-ql-select buffers-files query :action `(org-z--format-org-ql-heading ,window-width)))))
         (format-results (lambda (results)
                           (when results
                             (let* ((num-results (length results)))
                               (concat
                                (propertize
                                 (format "%d" num-results)
                                 'face 'bold)
                                (if (> num-results 1)
                                    " backlinks"
                                  " backlink"))))))
         (breadcrumb (org-eldoc-get-breadcrumb)))
    (message (funcall format-results results))
    (when breadcrumb
      (if (not results)
          breadcrumb
        (let* ((width (frame-width))
               (formatted-results (funcall format-results results))
               (results-len (length formatted-results))
               (breadcrumb-len (length breadcrumb))
               (total-len-w-space (+ results-len breadcrumb-len 3)))
          (cond
           ((and (<= total-len-w-space width)
                 (not (eq eldoc-echo-area-use-multiline-p t)))
            (concat breadcrumb " " "(" formatted-results ")"))
           ((eq eldoc-echo-area-use-multiline-p t)
            (apply
             #'concat
             breadcrumb
             (list
              "\n"
              formatted-results))
            )
           ((and (> total-len-w-space width)
                 (bound-and-true-p eldoc-echo-area-use-multiline-p))
            (apply
             #'concat
             breadcrumb
             (list
              "\n"
              formatted-results)))
           ((and (> total-len-w-space width)
                 (not (bound-and-true-p eldoc-echo-area-use-multiline-p)))
            (progn
              (setf (substring breadcrumb (- width (+ results-len 5))) (concat ".." " " "(" formatted-results ")"))
              breadcrumb))))))))

(defun org-z-eldoc-advice (orig-fun &rest args)
  (if (and org-z-mode org-z-show-backlinks)
      (or (org-z-eldoc-get-breadcrumb)
          (funcall orig-fun args))
    (funcall orig-fun args)))

;;;###autoload
(define-minor-mode org-z-mode
  "Minor mode for org-z."
  :lighter " org-z"
  :keymap (let ((map (make-sparse-keymap)))
            (define-key map (kbd "C-c C-.") 'org-z-insert-link)
            map)
  :group 'org-z
  :require 'org-z
  :global t
  (when (not org-id-link-to-org-use-id)
    (setq org-id-link-to-org-use-id t)))

(advice-add 'org-eldoc-documentation-function :around #'org-z-eldoc-advice)

(provide 'org-z)

;;; org-z.el ends here
